zetaclient/orchestrator/bootstrap.go

```diff
@@ -0,0 +1,414 @@
+	ctx context.Context,
+	client interfaces.ZetacoreClient,
+	tss interfaces.TSSSigner,
+	dbpath string,
+	logger base.Logger,
+	ts *metrics.TelemetryServer,
+) (map[int64]interfaces.ChainObserver, error) {
+	observerMap := make(map[int64]interfaces.ChainObserver)
+
+	_, _, err := syncObserverMap(ctx, client, tss, dbpath, logger, ts, &observerMap)
+	if err != nil {
+		return nil, err
+	}
+
+	return observerMap, nil
+}
+
+// syncObserverMap synchronizes the given observer map with the observers for all chains in the config.
+// This semantic is used to allow dynamic updates to the map.
+// Note (!) that it calls observer.Start() on creation and observer.Stop() on deletion.
+func syncObserverMap(
+	ctx context.Context,
+	client interfaces.ZetacoreClient,
+	tss interfaces.TSSSigner,
+	dbpath string,
+	logger base.Logger,
+	ts *metrics.TelemetryServer,
+	observerMap *map[int64]interfaces.ChainObserver,
+) (int, int, error) {
+	app, err := zctx.FromContext(ctx)
+	if err != nil {
+		return 0, 0, errors.Wrapf(err, "failed to get app context")
+	}
+
+	var (
+		added, removed int
+
+		presentChainIDs = make([]int64, 0)
+
+		onAfterAdd = func(_ int64, ob interfaces.ChainObserver) {
+			ob.Start(ctx)
+			added++
+		}
+
+		addObserver = func(chainID int64, ob interfaces.ChainObserver) {
+			mapSet[int64, interfaces.ChainObserver](observerMap, chainID, ob, onAfterAdd)
+		}
+
+		onBeforeRemove = func(_ int64, ob interfaces.ChainObserver) {
+			ob.Stop()
+			removed++
+		}
+	)
+
+	// EVM observers
+	for _, evmConfig := range app.Config().GetAllEVMConfigs() {
+		var (
+			chainID   = evmConfig.Chain.ChainId
+			chainName = evmConfig.Chain.ChainName.String()
+		)
+
+		if evmConfig.Chain.IsZetaChain() {
+			continue
+		}
+
+		chainParams, found := app.GetEVMChainParams(evmConfig.Chain.ChainId)
+		switch {
+		case !found:
+			logger.Std.Error().Msgf("Unable to find chain params for EVM chain %d", chainID)
+			continue
+		case !chainParams.IsSupported:
+			logger.Std.Error().Msgf("EVM chain %d is not supported", chainID)
+			continue
+		}
+
+		presentChainIDs = append(presentChainIDs, chainID)
+
+		// noop
+		if mapHas(observerMap, chainID) {
+			continue
+		}
+
+		// create EVM client
+		evmClient, err := ethclient.DialContext(ctx, evmConfig.Endpoint)
+		if err != nil {
+			logger.Std.Error().Err(err).Str("rpc.endpoint", evmConfig.Endpoint).Msgf("Unable to dial EVM RPC")
+			continue
+		}
+
+		database, err := db.NewFromSqlite(dbpath, chainName, true)
+		if err != nil {
+			logger.Std.Error().Err(err).Msgf("Unable to open a database for EVM chain %q", chainName)
+			continue
+		}
+
+		// create EVM chain observer
+		observer, err := evmobserver.NewObserver(
+			ctx,
+			evmConfig,
+			evmClient,
+			*chainParams,
+			client,
+			tss,
+			database,
+			logger,
+			ts,
+		)
+		if err != nil {
+			logger.Std.Error().Err(err).Msgf("NewObserver error for EVM chain %s", evmConfig.Chain.String())
+			continue
+		}
+
+		addObserver(chainID, observer)
```

Comment:
As mentioned in Slack. It's not easy to guarantee that the observer and signer are always added in pairs as we use separate function `syncObserverMap` and `syncSignerMap` to add them separately.

For incoming refactor, once we can rid of `GetAllEVMConfigs()`, `GetBTCChainAndConfig` etc. and use only one single map in the app context exposing two methods:
1. `GetEnabledExternalChains() []chains.Chain`
2. `GetExternalChainParams(chainID int64)`

we'll simply iterate all supported chains and get their corresponding chain params by a chain ID to create observer/signer in pairs. You can still refer to the implementation [here](https://github.com/zeta-chain/node/blob/orchestrator-add-new-chain-at-runtime/zetaclient/context/app_context.go#L123)

To make function shorter (now `syncObserverMap` spans 200+ lines and will increase if more chain comes in) and easier to unit test, you can create smaller functions like `syncObserverSignerMapEVM`, `syncObserverSignerMapBTC`, `syncObserverSignerMapSolana` to sync observer/signer pairs by consensus type such as `Consensus_ethereum`, `Consensus_bitcoin` and `Consensus_solana_consensus` etc.

Reply:
Yes, 100%

Reply:
https://github.com/zeta-chain/node/issues/2519